lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskStatus.cs

```diff
@@ -98,28 +98,48 @@ public void SetException(Exception e)
         {
             lock (_heartBeatManager)
             {
-                try
+                if (_lastException.IsPresent())
                 {
-                    if (!_lastException.IsPresent())
-                    {
-                        _lastException = Optional<Exception>.Of(e);
-                    }
+                    return;
+                }
 
-                    State = TaskState.Failed;
+                _lastException = Optional<Exception>.Of(e);
+                State = TaskState.Failed;
+
+                try
+                {
                     _taskLifeCycle.Stop();
                 }
-                finally
+                catch (Exception stopException)
                 {
-                    Heartbeat();
+                    Utilities.Diagnostics.Exceptions.Caught(
```

Comment:
Why not? We are talking about how to ensure the task thread is stopped when task stop handler is called, or task stop handler throw exception in SetException(). We should have a way to kill the task thread when exception happens and the task is still running. 
If this is no the right place to stop the task thread, then we can have another Jira to follow up it.